hpen.c

```diff
@@ -0,0 +1,491 @@
+	device_t sc_dev;
+
+	struct evdev_dev *sc_evdev;
+
+	uint8_t                 report_id;
+	bitstr_t                bit_decl(abs_caps, HPEN_N_USAGES_ABS);
+	bitstr_t                bit_decl(key_caps, HPEN_N_USAGES_KEY);
+	uint32_t                isize;
+	struct hpen_absinfo     ai[HPEN_N_USAGES_ABS];
+	struct hid_location     locs_abs[HPEN_N_USAGES_ABS];
+	struct hid_location     locs_key[HPEN_N_USAGES_KEY];
+};
+
+
+static device_probe_t hpen_probe;
+static device_attach_t hpen_attach;
+static device_detach_t hpen_detach;
+
+static evdev_open_t hpen_ev_open;
+static evdev_close_t hpen_ev_close;
+
+static const struct evdev_methods hpen_evdev_methods = {
+	.ev_open = &hpen_ev_open,
+	.ev_close = &hpen_ev_close,
+};
+
+static int
+hpen_ev_close(struct evdev_dev *evdev)
+{
+	device_t dev = evdev_get_softc(evdev);
+
+	mtx_assert(hidbus_get_lock(dev), MA_OWNED);
+
+	return (hidbus_set_xfer(dev, 0));
+}
+
+static int
+hpen_ev_open(struct evdev_dev *evdev)
+{
+	device_t dev = evdev_get_softc(evdev);
+
+	mtx_assert(hidbus_get_lock(dev), MA_OWNED);
+
+	return (hidbus_set_xfer(dev, HID_XFER_READ));
+}
+
+static void
+hpen_intr(void *context, void *buf, uint16_t len)
+{
+	device_t dev = context;
+	struct hpen_softc *sc = device_get_softc(dev);
+	size_t usage;
+	uint8_t id;
+
+	mtx_assert(hidbus_get_lock(dev), MA_OWNED);
+
+	/* Ignore irrelevant reports */
+	id = sc->report_id != 0 ? *(uint8_t *)buf : 0;
+	if (sc->report_id != id) {
+		DPRINTF("Skip report with unexpected ID: %hhu\n", id);
+		return;
+	}
+
+	/* Make sure we don't process old data */
+	if (len < sc->isize)
+		bzero((uint8_t *)buf + len, sc->isize - len);
+
+	/* Strip leading "report ID" byte */
+	if (sc->report_id) {
+		len--;
+		buf = (uint8_t *)buf + 1;
+	}
+
+	HPEN_FOREACH_USAGE_ABS(sc->abs_caps, usage) {
+		if (sc->locs_abs[usage].size > 0 && hpen_hid_map_abs[usage].code != HPEN_NO_CODE)
+			evdev_push_abs(sc->sc_evdev, hpen_hid_map_abs[usage].code,
+					hid_get_data(buf, len, &sc->locs_abs[usage]));
+	}
+
+	HPEN_FOREACH_USAGE_KEY(sc->key_caps, usage) {
+		if (sc->locs_key[usage].size > 0 && hpen_hid_map_key[usage].code != HPEN_NO_CODE)
+			evdev_push_key(sc->sc_evdev, hpen_hid_map_key[usage].code,
+					hid_get_data(buf, len, &sc->locs_key[usage]));
+	}
+
+	evdev_sync(sc->sc_evdev);
+}
+
+/* NOTE: in probe, called with sc==NULL just to check the pen collection exists */
+static int
+hpen_hid_parse(struct hpen_softc *sc, const void *d_ptr, uint16_t d_len,
+    uint32_t tlc_usage, uint8_t tlc_index)
+{
+	struct hid_item hi;
+	struct hid_data *hd;
+	size_t i;
+	bool pen_coll = false;
+	bool pen_found = false;
+
+	if (sc) {
+		bzero(sc->abs_caps, bitstr_size(HPEN_N_USAGES_ABS));
+		bzero(sc->key_caps, bitstr_size(HPEN_N_USAGES_KEY));
+	}
+
+	/* Parse inputs */
+	hd = hid_start_parse(d_ptr, d_len, 1 << hid_input);
+	HID_TLC_FOREACH_ITEM(hd, &hi, tlc_index) {
+		switch (hi.kind) {
+		case hid_collection:
+			if (hi.collevel == 1 && hi.usage ==
+			    HID_USAGE2(HUP_DIGITIZERS, HUD_PEN)) {
```

Comment:
No need for TLC usage check here. HID_TLC_FOREACH_ITEM macro does it better as it supports cases when several TLCs have the same usage.

Reply:
oh. This is leftover from before `HID_TLC_FOREACH_ITEM`, now I'm essentially duplicating the check that `HID_TLC_FOREACH_ITEM` already does..